Add article describing the delayed editor initialisation feature

sculpt0r opened the pull request:
Added a short article explaining in practice, what is the difference between two options in delayed editor creation.



Closes https://github.com/ckeditor/ckeditor4/issues/4553

f1ames commented on docs/features/delayed_creation/README.md:
```diff
@@ -0,0 +1,33 @@
+
+<info-box info="">This feature was introduced in 4.17.0 and by default it is off. It requires [enabling config option](../api/CKEDITOR_config.html#cfg-delayIfDetached) to worksfor the editor.</info-box>
+
+It was introduced because there were some issues during editor creation on the element that is detached from the DOM. During the entire process, there is some information extracted from the editor elements document. If the target element is detached, it is simply impossible to access them. This situation may be caused unintentionally by using CKEditor 4 with popular frameworks (e.g. [Angular inte ...
```
I would start by saying how it works / what it does. So that "it prevents editor creation when parent element is detached from DOM...".

Then maybe a separate section on possible issue it can solve (for example using CKEditor 4 in modals, using with integrations and so on). So the similar description you had already added. This can give broader context and clarity when to use it.

f1ames replied:
One more thing, apart from using "some" like:

> there were some issues during editor

or

> During the entire process, there is some information extracted from

You may use more specific examples, or even link to GH issues. So it would be better to describe specific cases more precisely or use term like "specific issues", "group of issues", e.g.:

> It was introduced because we have observed a group of issues when editor initialization failed during editor creation on the element that is detached from the DOM.

sculpt0r replied:
> You may use more specific examples, or even link to GH issues.

I had doubts about linking to GH issues in documentation. But if you say we can do this... 👍

sculpt0r replied:
And finally I decide to manage that inside the doc article ;)

f1ames commented on docs/features/delayed_creation/README.md:
```diff
@@ -0,0 +1,33 @@
+For licensing, see LICENSE.md.
+-->
+
+# Delayed creation
```
```suggestion
# Delayed editor creation
```
?

f1ames commented on docs/features/delayed_creation/README.md:
```diff
@@ -0,0 +1,33 @@
+
+## Two ways to delay creation
+
+- Enabling [delayIfDetached option](../api/CKEDITOR_config.html#cfg-delayIfDetached) enabled the default interval checks. A check is performed every [amount of time](../api/CKEDITOR_config.html#cfg-delayIfDetached_interval). Whenever a target element is found to be attached to the DOM - the new instance of the editor is created. This is the easiest and fastest way to use this feature. It may be n ...
```
```suggestion
- The [delayIfDetached](../api/CKEDITOR_config.html#cfg-delayIfDetached) config option which is enabled by default results in editor element being checked in given intervals. Checks are performed every given amount of time (which can be changed via [delayIfDetached_interval](../api/CKEDITOR_config.html#cfg-delayIfDetached_interval)). Whenever a target element is found to be attached to the DOM - the editor instance is then created. For more advanced usage refer to the `callback` config option described below.
```

f1ames commented on docs/features/delayed_creation/README.md:
```diff
@@ -0,0 +1,33 @@
+
+- Enabling [delayIfDetached option](../api/CKEDITOR_config.html#cfg-delayIfDetached) enabled the default interval checks. A check is performed every [amount of time](../api/CKEDITOR_config.html#cfg-delayIfDetached_interval). Whenever a target element is found to be attached to the DOM - the new instance of the editor is created. This is the easiest and fastest way to use this feature. It may be n ...
+
+- Additional setting [callback](../api/CKEDITOR_config.html#cfg-delayIfDetached_callback) will turn off the default interval checks. Instead, the callback is invoked with a single argument. This argument is a function that should be invoked to finish editor creation. The advantage here is that it gives full control over the time that the actual creation is performed. A received callback may be st ...
```
```suggestion
- To have more control over when editor instance will be initialized the [delayIfDetached_callback](../api/CKEDITOR_config.html#cfg-delayIfDetached_callback) config option is provided. When set as `function`, it will be used instead of interval checks and invoked with a single argument. This argument is a function that should be invoked to finish editor creation. The advantage here is that it gives full control over the time that the actual creation is performed. A received callback may be stored and used right after the target element is attached to DOM.
```

f1ames commented on docs/features/delayed_creation/README.md:
```diff
@@ -0,0 +1,33 @@
+
+## Practical example
+
+Let's have a target element and it is not visible by the user (is detached), but the editor creation was called.
+
+Whenever user clicks the button: The element is visible again.
+
+With the interval checks, depends on the use case, the script may run for so long until the first button use. In the worst case, the button may not be clicked at all. Also, last check may happen milliseconds before user clicks. The default timeout is small enough, that there will be no flickering effect, but still, we need to wait that amount of time and then the creation time.
+
+With the callback option, we may store the received creation function. Bind to the button onclick and invoke the stored function. Now we need to wait only for creation time. No additional delays or unnecessary checks. However, additional logic needs to be implemented and binded to every place that could show the editor.
```
Maybe it can be described in form of the list like elements:

> Some intro describing initial situation...
> 1. User clicks a button.
> 1. Editor parent element becomes visible.
> 1. ...

sculpt0r replied:
I decided to describe it is a little bit different approach. Pleas take a look at the summary comment:

f1ames commented on docs/features/delayed_creation/README.md:
```diff
@@ -0,0 +1,33 @@
+order: 60
+url: features/delayed_creation
+menu-title: Delayed creation
+meta-title-shor: Delayed creation
```
```suggestion
meta-title-short: Delayed creation
```
? :thinking:

f1ames reviewed (changes requested):
I like the idea of the separate section and the broad description since the concept described here is not that easy to understand :+1: Still, I would try to focus on making the description as simple as possible (which doesn't mean short) to make it easy to understand. I added few suggestions and corrections, but feel free to adjust them further.

I also wonder if we should use term like "editor (parent) element" or maybe "editor container"?

Hmmm, we have both Guides and Features section which is sometimes confusing. Guides are kind of dev guides so the topic of delaying editor (as this is configuration stuff and may require coding) fits there. OTOH it's a feature which can be simply enabled via config so fits in Features section too :thinking: 
I guess it will be good to have discoverability in mind (how the person not knowing about this feature can find it).

Apart from that I see common cases where it is useful are already mentioned but it will be good to mention cases where it might break something (e.g. editor not initializing) so one may try to disable this feature to solve the issue.

sculpt0r commented:
> I also wonder if we should use term like "editor (parent) element" or maybe "editor container"?

I thought about that. Probably I should check if I'm 100% consistent here, but:
- I prefer "editor parent element" when we talk about detaching elements. Because not only the closest parent may be detached, but any parent element. And by "editor container" ([even according to our API](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#property-container)) we are calling the closest parent. 🤔

sculpt0r commented:
Hi @f1ames, ready for another review. The changes:
- I correct **some** ;p;p of the paragraphs
- Added a troubleshooting paragraph - which should exactly show what was the problem with the detached element. And finally, I resign from linking to any GH issue.
- I separate almost every paragraph on two sections: the `interval` and the `callback` - which should better visualize differences between those two approaches.
- Added a paragraph how to grab the editor reference when this feature is usage.

f1ames commented on docs/features/delayed_creation/README.md:
````diff
@@ -0,0 +1,130 @@
+
+You can get an instance in different ways, depending on used options:
+
+### Interval approach
+```js
+var editor = CKEDITOR.instances[ 'editorName' ];
+```
````
I would mention here that getting editor ref should be done on `instanceReady` event. Because when this code is called before editor is created it will fail AFAIR.

sculpt0r replied:
Well, you can grab an instance like that, but need to wait for `instanceReady` anyway to use API. Here, I just want to point out that `the old` way of creating an editor may return `null` whereas the object might be returned in the old code.

But, yes - I will mention that :)

f1ames reviewed (changes requested):
Looks good, I did some corrections here and there - could you take a look?

I still have a feeling that we are missing basic, general overview what this feature does. It is easy to understand for us, but when reading docs it seems we go right away into how to use it and approaches there are. But the introduction is a bit laconic :thinking: WDYT? Maybe we could expand it somehow?

f1ames commented:
One more thing I forgot to mention during review :see_no_evil: This PR relates to changes in the upcoming CKEditor 4 major so should target `major` branch.

sculpt0r commented:
> One more thing I forgot to mention during review 🙈 This PR relates to changes in the upcoming CKEditor 4 major so should target `major` branch.

The rebasing was kind of painful... So I decided to create another PR but saving commits from this one: https://github.com/ckeditor/ckeditor4-docs/pull/364. Sorry about that.
